libs/datamodel/core/src/transform/ast_to_dml/db/relations.rs

```diff
@@ -0,0 +1,219 @@
+
+    RelationEvidence {
+        ast_model,
+        model_id: *model_id,
+        ast_field,
+        field_id: *field_id,
+        relation_field,
+        opposite_model,
+        is_self_relation,
+        opposite_relation_field,
+    }
+}
+
+fn ingest_relation<'ast, 'db>(evidence: RelationEvidence<'ast, 'db>, relations: &mut Relations<'ast>) {
+    let relation_type = match (evidence.ast_field.arity, evidence.opposite_relation_field) {
+        (ast::FieldArity::List, Some((opp_field_id, opp_field, _))) if opp_field.arity.is_list() => {
+            // This is an implicit many-to-many relation.
+
+            // We will meet the relation twice when we walk over all relation
+            // fields, so we only instantiate it when the relation field is that
+            // of model A, and the opposite is model B.
+            if evidence.ast_model.name.name > evidence.opposite_model.name.name {
+                return;
+            }
+
+            // For self-relations, the ordering logic is different: model A
+            // and model B are the same. The _first field in source text
+            // order_ is field A, the second field is field B.
+            if evidence.is_self_relation && evidence.ast_field.name.name > opp_field.name.name {
+                return;
+            }
+
+            RelationType::ImplicitManyToMany {
+                field_a: evidence.field_id,
+                field_b: opp_field_id,
+            }
+        }
+        (ast::FieldArity::List, Some(_)) => {
+            // This is a 1:m relation defined on both sides. We skip the virtual side.
+            return;
+        }
+        (ast::FieldArity::List, None) => {
+            // This is a 1:m relation defined on the virtual side only.
+
+            RelationType::OneToMany(OneToManyRelationFields::Back(evidence.field_id))
+        }
+        (ast::FieldArity::Optional | ast::FieldArity::Required, Some((opp_field_id, opp_field, _)))
+            if opp_field.arity.is_required() || opp_field.arity.is_optional() =>
+        {
+            // This is a 1:1 relation.
+            RelationType::OneToOne {
+                field_a: evidence.field_id,
+                field_b: opp_field_id,
+            }
+        }
+        (ast::FieldArity::Optional | ast::FieldArity::Required, None) => {
+            // This is a 1:m relation defined on the concrete side only.
+            RelationType::OneToMany(OneToManyRelationFields::Forward(evidence.field_id))
+        }
+        (ast::FieldArity::Optional | ast::FieldArity::Required, Some((opp_field_id, _, _))) => {
+            // This is a 1:m relation defined on both sides.
+            RelationType::OneToMany(OneToManyRelationFields::Both(evidence.field_id, opp_field_id))
+        }
+    };
+
+    let relation = Relation {
+        tpe: relation_type,
+        relation_name: evidence.relation_field.name,
+    };
+
+    let relation_idx = relations.relations_storage.len();
```

Comment:
Indentation for the whole end of the file, nitpick schnitpick.

Reply:
You mean blank lines between statements?

Reply:
Will do.

Reply:
Ah, yes, that's it.